### DISCUSSION TITLE
When using machine learning models, how can I use my own INDICATOR data that I have already computed, instead of computing it in real time？

### DISCUSSION THREAD

**Discussion by yys123** [edtechre/pybroker#15] [Q&A] [answered]
When using machine learning models, how can I use my own indicator data that I have already computed, instead of computing it in pybroker？I want to use dataframe form to import indicator data directly.

**Comment by edtechre** [comment, marked as answer]
Hi @yys123,

You can add the indicator data as columns to your DataFrame and register them using [pybroker.register_columns()](https://www.pybroker.com/en/latest/reference/pybroker.scope.html#pybroker.scope.register_columns). For an example, refer to the ``rsi`` column in the [Creating a Custom Data Source](https://www.pybroker.com/en/latest/notebooks/7.%20Creating%20a%20Custom%20Data%20Source.html) notebook.

**Reply by yys123** [reply]
**Thank you for your reply.
I have done as you said, if I run this code,  it has no problem:**

_config = StrategyConfig(bootstrap_sample_size=100)
strategy = Strategy(YFinance(), '3/1/2017', '3/1/2022', config)
strategy.add_execution(**None**, ['NVDA', 'AMD'], models=model_slr)
strategy.backtest(train_size=0.5)_

**but when running the following code, it reports an error:**

_config = StrategyConfig(bootstrap_sample_size=100)
strategy = Strategy(YFinance(), '3/1/2017', '3/1/2022', config)
strategy.add_execution(**hold_long**, ['NVDA', 'AMD'], models=model_slr)
strategy.backtest(train_size=0.5)_
**error：**
### ValueError: at least one array or dtype is required

### Besides, the 'strategy.walkforward' function can not work in any case if I used my own indicator data.

**Reply by edtechre** [reply]
The error you pasted sounds like it's coming from Scikit learn. Normally, indicators computed in PyBroker are used as columns for the input data passed to model#predict(). Since you are using custom columns, you will need to set [input_data_fn](https://www.pybroker.com/en/latest/reference/pybroker.model.html#pybroker.model.model) to pass your custom columns to your model when making predictions in ``hold_long`` e.g.

```
pybroker.register_columns('ind1', 'ind2', 'ind3')

def input_data_fn(df):
    return df[['ind1', 'ind2', 'ind3']]  # Passes ind1, ind2, ind3 columns to model#predict()

model = pybroker.model('my_model', train_fn, input_data_fn=input_data_fn)
```
You were seeing the error above only for ``hold_long`` because predictions are only computed when there is an execution function set.

`Besides, the 'strategy.walkforward' function can not work in any case if I used my own indicator data.`

Calling ``strategy.walkforward`` will work with your own indicator data by making the change I described.

**Reply by yys123** [reply]
Thank you for your answer, it helped me solve this problem!

**Reply by edtechre** [reply]
Note, as of v1.1.12, you can now return a list of column names to use as input from your train function:

```
def train_fn(symbol, train_data, test_data):
    ...
    return model, ['ind1', 'ind2', 'ind3']
```

You will still need to register ind1, ind2, ind3 using ```pybroker.register_columns```.
